sotodlib/tod_ops/jumpfind.py

```diff
@@ -0,0 +1,406 @@
+
+    Arguments:
+
+        x: Data to jumpfind on, expects 1D
+
+        sensitivity: Sensitivity of the jumpfinder, roughly correlates with
+                     1/(jump size) but since data is filtered, detrended, and
+                     rescaled during jumpfinding it is better to think of it as
+                     something non-physical.
+
+        max_depth: The maximum recursion depth, set negetive to not use
+
+        depth: The current recursion depth
+
+    Returns:
+
+        jumps: The indices of jumps in x
+               There is some uncertainty on order of a few samples
+               Jumps within 20 samples of each other may not be distinguished
+    """
+    _x = sig.detrend(x)
+    if _x.std() > 10:
+        _x = _x / (10 ** (int(np.log10(_x.std()))))
+    if np.isclose(_x.std(), 0.0):
+        return np.array([])
+
+    jumps = jumpfind_gaussian(
+        _x,
+        2 * _x.std(),
+        0,
+        10,
+        1.0 / sensitivity,
+        0,
+        20,
+        0,
+        height=1,
+        prominence=1,
+    )
+
+    # If no jumps found return
+    if len(jumps) == 0:
+        return jumps
+
+    # If at max_depth then return
+    if depth == max_depth:
+        return jumps
+
+    # Recursively check for jumps between jumps
+    _jumps = np.insert(jumps, 0, 0)
+    _jumps = np.append(_jumps, len(x))
+    added = 0
+    for i in range(len(_jumps) - 1):
+        sub_jumps = jumpfind_recursive_gaussian(
+            x[(_jumps[i]) : (_jumps[i + 1])], sensitivity, max_depth, depth + 1
+        )
+        jumps = np.insert(jumps, i + added, sub_jumps + _jumps[i])
+        added += len(sub_jumps)
+    return jumps
+
+
+def jumpfind_sliding_window(
+    x, window_size=10000, overlap=1000, jumpfinder=jumpfind_tv, **kwargs
+):
+    """
+    Run jumpfinder through a sliding window.
+    This can help get jumps towards the edges of the data that may be missed.
+    Nominally those jumps can be found if the jumpfinder reaches sufficient depth,
+    but sometimes it takes tweaking of the parameters to catch them.
+
+    Arguments:
+
+        x: Data to jumpfind on, expects 1D
+
+        window_size: Size of window to use
+
+        overlap: Overlap between adjacent windows
+
+        jumpfinder: Jumpfinding function to use
+
+        **kwargs: Keyword args to pass to jumpfinder
+
+    Returns:
+
+        jumps: The indices of jumps in x
+               There is some uncertainty on order of 1 sample
+               Jumps within min_chunk of each other may not be distinguished
+    """
+    jumps = np.array([], dtype=int)
+    for i in range(len(x) // (window_size - overlap)):
+        start = i * (window_size - overlap)
+        end = np.min((start + window_size, len(x)))
+        _jumps = jumpfinder(x[start:end], **kwargs) + start
+        jumps = np.hstack((jumps, _jumps))
+    return np.unique(jumps).astype(int)
+
+
+def jumpfind(tod, signal=None, buff_size=10, jumpfinder=jumpfind_tv, **kwargs):
```

Comment:
It looks like this is supposed to be the main user interface function.  But using it is awkward, because the user needs to specify a particular jumpfinder to use, and then also needs to look up the specific kwargs that that jumpfinder accepts, and tune the parameters for their case of interest.  The interface for jumpfinder functions is totally free right now (it's not even true that they all accept the "min_size" parameter), so it's not possible to do simple experiments where you individually adjust parameters in the jumpfind call (such as jumpfinder=...) to compare results from different jumpfinders.

This means there's not much value added by this version jumpfind function -- in fact it's almost more work to use jumpfind than to use a jumpfinder directly.

I think you should either remove this main entry point, and have the 2-D support & RangesMatrix conversion supported in each individual jumpfinder OR else make jumpfind a little more powerful and general:

* Allow user to pass jumpfinder function by string name (e.g. "tv")
* Have the function accept some basic parameters that control the jumpfinders.  For example, "min_size" is always something the user should be able to specify.  I think you should also have a "min_sigma", which will lead to a min_size computed based on the stdev (appropriately estimated).  Another basic parameter would be the time scale for a jump (how many samples does a particular jump take to occur -- this can used to tease out true glitchy jumps from bright signals that are a little slower, such as planet signals; in code the time scale would be converted to a sample count and used to set windowing and filtering parameters).
* In the case that jumpfinder is specified by name, compute reasonable default parameters to pass to that jumpfinder, _that are based on the data_ and those general basic parameters.  "Based on the data" just means that it can figure out stdev on its own, it doesn't need the user to pass in stdev or min_size all the time.  The user can override the jumpfinder arguments using kwargs ... but jumpfind should provide sensible defaults.

In simple cases, I should get more or less the same answer.  E.g. with just large steps in white noise, these two should both work and give me the ~same answers:
```
  j1 = jumpfind(tod, jumpfinder='tv')
  j2 = jumpfind(tod, jumpfinder='gaussian')
```

Reply:
Yeah. The main reason why I resisted having each function support 2d and ranges previously was because when testing its was more convenient to avoid all of that (and axis managers) so that I could quickly make some noise + jumps and check how well the output array of jumps matched the input. So maybe the correct thing would be to do make them all support it. I do like having some sort of wrapper for the other reason of it avoiding having to copy the same sort of code at the start and end of every function. But I do agree it is confusing to use if you don't already know all the kwargs for the jumpfinders (which for obvious reasons was not an issue for me).

Making it more powerful and general sounds good but I'll have to think a bit on how to deal with the differing kwargs in a way thats not just as annoying. Maybe I want to add more power to the wrapper and then just add an additional function for each jumpfinder that calls the wrapper but with the jwargs specified? It adds function bloat to the file but then at least the arguments will show up when people `?` the functions in an interactive shell/notebook and show up in autocomplete for people whose editors are configured for that.

> Allow user to pass jumpfinder function by string name (e.g. "tv")

I have mixed feelings on this. I understand the convenience from a user perspective, but these sorts of "select the mode with a string" parameters always add the need for a set of if statements at the start of the function that needs to be changed whenever a mode is added/deleted which feels clunky to me vs just passing a function.

> Have the function accept some basic parameters that control the jumpfinders. For example, "min_size" is always something the user should be able to specify. 

This makes sense.

> I think you should also have a "min_sigma", which will lead to a min_size computed based on the stdev (appropriately estimated). 

The problem here is that std is effected by jumps. So if you do this and have big jumps it can skew the std enough to drop even mid sized jumps below your threshold. One way around this (which one of the jumpfinder tries to do), is to recompute the std and the stuff based on it at each recursion, the downside of this is it tends to find less jumps per recursion to it takes more iterations to complete. 

> Another basic parameter would be the time scale for a jump (how many samples does a particular jump take to occur -- this can used to tease out true glitchy jumps from bright signals that are a little slower, such as planet signals; in code the time scale would be converted to a sample count and used to set windowing and filtering parameters).

Yeah, this is something I have toyed around with in the past. I didn't find a great way to determine the time scale of the jump. Part of what complicates this is that when you zoom in on them a good chunk of jumps are actually multiple jumps happening in quick succession. But its also not something I spent a ton of time on so I can certainly revisit it. 

> In the case that jumpfinder is specified by name, compute reasonable default parameters to pass to that jumpfinder, that are based on the data and those general basic parameters. "Based on the data" just means that it can figure out stdev on its own, it doesn't need the user to pass in stdev or min_size all the time. The user can override the jumpfinder arguments using kwargs ... but jumpfind should provide sensible defaults.

This also seems reasonable. I have a function that tries to do this for the gaussian jumpfinder that typically works better than me specifying params by default. But the issue with this (as I stated above) is the way that jumps skew the std.

>  In simple cases, I should get more or less the same answer. E.g. with just large steps in white noise, these two should both work and give me the ~same answers:

Yeah, as far as I can tell this is true. You typically only see the difference in their ability to (not) flag sources and in more jumpy regions. I actually don't know if there is a case where the gaussian one outperforms the TV one, I just left it in just in case we do find such a case down the line.

Reply:
> > I think you should also have a "min_sigma", which will lead to a min_size computed based on the stdev (appropriately estimated).
> 
> The problem here is that std is effected by jumps. So if you do this and have big jumps it can skew the std enough to drop even mid sized jumps below your threshold. One way around this (which one of the jumpfinder tries to do), is to recompute the std and the stuff based on it at each recursion, the downside of this is it tends to find less jumps per recursion to it takes more iterations to complete.

Yes, "appropriately estimated" was meant to get at that.  Here's a trick I often use to get a white noise stdev estimate, in the presence of 1/f and jumps/glitches:
```
# Get difference vector.
delta = np.diff(signal)
# Find ~1 sigma limits
lims = np.quantiles(delta, [0.159, 0.841])
# Convert to equivalent signal stdev
sigma = (lims[1] - lims[0]) / 8**.5
```
That's robust to outliers.

Reply:
Ah, that's clever. I'll give it a shot. Seems like something that should maybe exist as a function somewhere in sotodlib, this is likely not the only place where a robust estimate of the noise level (but not a full calculation of the white noise level) would be useful.

Reply:
@mhasself would you mind taking a look at some of the changes I've made to this in terms of making it more general. I've only done a bit but I want to make sure that I'm moving in a reasonable direction (I always have a hard time judging how bad my interfaces are since having written them using them is easy for me).

I'll throw in some stuff with determining parmeters based on stdev and min_sigma and whatnot in a bit. The initial testing I have done with them have gone well so I don't foresee any problems there.

Reply:
Yes, I like the changes!

Couple of notes as you continue ... :  I think it's considered good practice to pass parameters in key=value form, especially when there's a chance that the interface here will be changed or extended.  So instead of:
```
           jumps = jumpfinder(
                _signal, min_chunk, min_size, win_size, max_depth, **kwargs
            )
```
you should probably use
```
           jumps = jumpfinder(
                _signal, min_chunk=min_chunk, min_size=min_size, win_size=win_size, max_depth=max_depth, **kwargs
            )
```

Also, you say in a comment that kwargs have to be passed in a certain order ... they do not, because they are kwargs.  The function will pull out any keys it recognizes, and leave the rest in its own kwargs.

Reply:
> Couple of notes as you continue ... : I think it's considered good practice to pass parameters in key=value form, especially when there's a chance that the interface here will be changed or extended. So instead of:

Sure. 

> Also, you say in a comment that kwargs have to be passed in a certain order ... they do not, because they are kwargs. The function will pull out any keys it recognizes, and leave the rest in its own kwargs.

Ah that's right. For some reason I was under the impression that if you pass one that it doesn't recognize first it will assume the ones that come after are also part of the kwargs. But I guess its a dict so the ordering doesn't matter.